File: exercises/fila/Fila.py
from fila_exceptions import FilaCheiaException, FilaVaziaException

class Fila:
    def __init__(self, size=100):
        self._fila = []
        self._inicio = 0
        self._fim = 0
        self._tamanho = size
        self._n_elementos = 0

    def __str__(self):
        return "%s" % self._fila

    @property
    def primeiro(self):
        return self._fila[self._inicio]

    @property
    def vazia(self):
        return self._n_elementos == 0

    @property
    def cheia(self):
        return self._n_elementos == self._tamanho

    def entrar(self, valor):
        if self.cheia:
            raise FilaCheiaException('Fila cheia!')

        self._fila[self._fim] = valor
        self._fim = (self._fim + 1) % self._tamanho
        self._n_elementos += 1

    def sair(self):
        if self.vazia:
            raise FilaVaziaException('Fila vazia!')

        valor = self._fila[self._inicio]
        self._inicio = (self._inicio + 1) % self._tamanho
        self._n_elementos -= 1

        return valor

File: exercises/fila/Fila.java
public class Fila {
  private int tamanho;
  private int nElementos = 0;
  private int inicio = 0;
  private int fim = 0;
  private int[] fila;

  public Fila(int tamanho) {
    this.tamanho = tamanho;
    this.fila = new int[tamanho];
  }

  public boolean cheia() {
    return this.nElementos == tamanho;
  }

  public boolean vazia() {
    return this.nElementos == 0;
  }

  public int sair() throws Exception {
    if (this.vazia()) {
      throw new Exception("Fila vazia");
    }

    int valor = this.fila[this.inicio];
    this.inicio = (this.inicio + 1) % this.tamanho;
    this.nElementos--;

    return valor;
  }

  public void entrar(int valor) throws Exception {
    if (this.cheia()) {
      throw new Exception("Fila cheia");
    }

    this.fila[this.fim] = valor;
    this.fim = (this.fim + 1) % this.tamanho;
    this.nElementos++;
  }
}
